core/local/local.go

```diff
@@ -23,530 +23,367 @@ package local
-	e.stages = s
-}
-
-func (e *Executor) GetIterations() int64 {
-	return atomic.LoadInt64(&e.iters)
-}
-
-func (e *Executor) GetEndIterations() null.Int {
-	v := atomic.LoadInt64(&e.endIters)
-	if v < 0 {
-		return null.Int{}
-	}
-	return null.IntFrom(v)
-}
+// Run the ExecutionScheduler, funneling all generated metric samples through the supplied
+// out channel.
+func (e *ExecutionScheduler) Run(ctx context.Context, engineOut chan<- stats.SampleContainer) error {
+	executorsCount := len(e.executors)
+	logger := e.logger.WithField("phase", "local-execution-scheduler-run")
+	e.initProgress.Modify(pb.WithConstLeft("Run"))
 
-func (e *Executor) SetEndIterations(i null.Int) {
-	if !i.Valid {
-		i.Int64 = -1
+	if e.state.IsPaused() {
+		logger.Debug("Execution is paused, waiting for resume or interrupt...")
+		e.initProgress.Modify(pb.WithConstProgress(1, "paused"))
+		select {
+		case <-e.state.ResumeNotify():
+			// continue
+		case <-ctx.Done():
+			return nil
+		}
 	}
-	e.Logger.WithField("i", i.Int64).Debug("Local: Setting end iterations")
-	atomic.StoreInt64(&e.endIters, i.Int64)
-}
-
-func (e *Executor) GetTime() time.Duration {
-	return time.Duration(atomic.LoadInt64(&e.time))
-}
 
-func (e *Executor) GetEndTime() types.NullDuration {
-	v := atomic.LoadInt64(&e.endTime)
-	if v < 0 {
-		return types.NullDuration{}
-	}
-	return types.NullDurationFrom(time.Duration(v))
-}
+	e.state.MarkStarted()
+	defer e.state.MarkEnded()
+	e.initProgress.Modify(pb.WithConstProgress(1, "running"))
 
-func (e *Executor) SetEndTime(t types.NullDuration) {
-	if !t.Valid {
-		t.Duration = -1
-	}
-	e.Logger.WithField("d", t.Duration).Debug("Local: Setting end time")
-	atomic.StoreInt64(&e.endTime, int64(t.Duration))
-}
+	logger.WithFields(logrus.Fields{"executorsCount": executorsCount}).Debugf("Start of test run")
 
-func (e *Executor) IsPaused() bool {
-	e.pauseLock.RLock()
-	defer e.pauseLock.RUnlock()
-	return e.pause != nil
-}
+	runResults := make(chan error, executorsCount) // nil values are successful runs
 
-func (e *Executor) SetPaused(paused bool) {
-	e.Logger.WithField("paused", paused).Debug("Local: Setting paused")
-	e.pauseLock.Lock()
-	defer e.pauseLock.Unlock()
+	runCtx, cancel := context.WithCancel(ctx)
+	defer cancel() // just in case, and to shut up go vet...
 
-	if paused && e.pause == nil {
-		e.pause = make(chan interface{})
-	} else if !paused && e.pause != nil {
-		close(e.pause)
-		e.pause = nil
+	// Run setup() before any executors, if it's not disabled
+	if !e.options.NoSetup.Bool {
+		logger.Debug("Running setup()")
+		e.initProgress.Modify(pb.WithConstProgress(1, "setup()"))
+		if err := e.runner.Setup(runCtx, engineOut); err != nil {
+			logger.WithField("error", err).Debug("setup() aborted by error")
+			return err
+		}
 	}
-}
-
-func (e *Executor) GetVUs() int64 {
-	return atomic.LoadInt64(&e.numVUs)
-}
+	e.initProgress.Modify(pb.WithHijack(e.getRunStats))
+
+	runCtxDone := runCtx.Done()
+	runExecutor := func(executor lib.Executor) {
```

Comment:
Minor: could you extract this into a standalone private method? It would be easier to test that way.

Reply:
Resolved by https://github.com/loadimpact/k6/pull/1007/commits/a3da2afd1ff4d8592f12040750e69498f5fc015a